Add reader and writer for Puffin, indexes and stats file format

findepi opened the pull request:
Format documentation: https://github.com/apache/iceberg/pull/4944

findepi commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,136 @@
+  static final int SUPPORTED_FLAGS = 0b1;
+  static final int FLAG_COMPRESSED = 0b1;
+
+  // TODO use LZ4 with frames, see https://trinodb.slack.com/archives/CP1MUNEUX/p1649676596198729
```
This needs to be addressed before the PR is merged.

rdblue replied:
And that should be documented in the spec, right?

findepi replied:
LZ4 frame support draft in aircompressor: https://github.com/airlift/aircompressor/pull/142

rdblue commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,69 @@
+  private final long length;
+  private final Optional<String> compressionCodec;
+
+  @JsonCreator
```
Because Iceberg is a library, we can't use annotations to control serialization. The JVM makes no guarantee that annotations are available at runtime and users are in control of the final classpath. So this brings in a new set of correctness problems. That's why we always serialize with a `SomethingParser` class.

rdblue replied:
No worries! This is a problem most people aren't aware of so it's pretty common.

rdblue commented on build.gradle:
```diff
@@ -217,9 +217,11 @@ project(':iceberg-core') {
     implementation 'org.apache.httpcomponents.client5:httpclient5'
     implementation "com.fasterxml.jackson.core:jackson-databind"
     implementation "com.fasterxml.jackson.core:jackson-core"
+    implementation "com.fasterxml.jackson.datatype:jackson-datatype-jdk8"
```
This shouldn't be needed.

findepi replied:
`jackson-datatype-jdk8` no longer used here

rdblue commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,69 @@
+    this.compressionCodec = Objects.requireNonNull(compressionCodec, "compressionCodec is null");
+  }
+
+  public String getType() {
```
Iceberg avoids the use of `get` in method names. In cases like this one, it's an unnecessary additional word. And in most other cases, there's a more specific verb that should be used instead, like `find` or `fetch` that give the reader more context about what's happening.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,136 @@
+
+  static final int CURRENT_FORMAT_VERSION = 1;
+
+  private static final byte[] MAGIC = new Base16().decode("50464953");
```
Why not just include the values from the spec here?

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,136 @@
+        (Byte.toUnsignedInt(data[offset + 3]) << 24);
+  }
+
+  static byte[] compressFooterPayload(byte[] payload) {
```
We don't generally work with `byte[]`, especially with no offset or length, because it requires new allocations all the time. Can these work with `ByteBuffer` instead?

findepi replied:
This will require https://github.com/airlift/aircompressor/pull/142#discussion_r848749110.
To unblock this, I will drop LZ4 for now from this PR.

findepi commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,69 @@
+    return length;
+  }
+
+  public Optional<String> getCompressionCodec() {
```
i would ask the question the other way around, but fine, changed to nullable.

findepi commented:
@rdblue i applied most of the comments (no more `jackson-datatype-jdk8`, json-related annotations, `get` in accessor names). I also switched over from base16-encoded test "fixtures" to test resources.

I switched writing to use `ByteBuffer`, this removes removed data copying upon compression.
This is as a fixup for now. Please let me know if you want me to use `ByteBuffer` in the reader/writer APIs as well?
It seems `byte[]` works there well (eg because uncompressed size is known, so no need to re-allocation), so please confirm before I apply further changes.

rdblue commented:
@findepi, I would prefer to use `ByteBuffer` everywhere. If the backing buffer was easy to allocate and work with, that's great. But you're not forced to reallocate if it is not.

singhpk234 commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,147 @@
+
+  public void finish() throws IOException {
+    writeHeaderIfNeeded();
+    if (finished) {
+      throw new IllegalStateException("Already finished");
+    }
```
can call `checkNotFinished()`

findepi replied:
done

findepi commented:
( squashed and rebased on top of current version of https://github.com/apache/iceberg/pull/4534, no other changes in this push )

findepi commented on core/src/main/java/org/apache/iceberg/stats/FileMetadataParser.java:
```diff
@@ -0,0 +1,147 @@
+    for (Map.Entry<String, String> entry : fileMetadata.properties().entrySet()) {
+      generator.writeStringField(entry.getKey(), entry.getValue());
+    }
+    generator.writeEndObject();
+
+    generator.writeEndObject();
```
first ends `properties` field, the second end the whole `FileMetadata`.

findepi commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,122 @@
+    switch (codec) {
+      case LZ4: {
+        // TODO requires LZ4 frame decompressor, e.g. https://github.com/airlift/aircompressor/pull/142
+        throw new UnsupportedOperationException("LZ4 is not supported yet");
```
thanks for the tip

findepi commented:
AC 
@rdblue  @singhpk234 thanks for your review, let me know what else I can change.

kbendick commented on core/src/main/java/org/apache/iceberg/stats/StatsCompressionCodec.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  public static StatsCompressionCodec forName(String codecName) {
+    Objects.requireNonNull(codecName, "codecName is null");
```
Nit: Usually we use the relocated `Preconditions` form.

Also, does it make more sense to just maintain a mapping between the two vs iterating over them?

findepi replied:
from perf perspective, for enum with two elements, i don't know which is actually faster
from readability perspective, it's probably subjective which is more readable, so which one you'd prefer?

kbendick replied:
For 2 elements actually it’s probably overkill in my opinion. Good point.

kbendick commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+  static final int FLAG_COMPRESSED = 0b1;
+
+  static byte[] getMagic() {
+    return new byte[] {0x50, 0x46, 0x49, 0x53};
```
Question: Does this particular magic bytes correlate with anything in particular or did we just decide on it?

findepi replied:
This is from the format spec https://github.com/apache/iceberg-docs/pull/69:

> `Magic` is four bytes 0x50, 0x46, 0x49, 0x53 (short for: Plain Format for
  Indices and Statistics),

RussellSpitzer replied:
This always allocates bytes for the Magic byte[] array, Is this because we want to re-use the array we are allocating here? Just wondering if this could be a static final byte array since it get's used in some contexts (like counting length) which don't require the allocation.

rdblue replied:
While it is mutable, it should be fine to rely on people not modifying it. This isn't public.

findepi replied:
I don't see a benefit of declaring it as a constant field, and i do feel strongly about sharing mutable state that's not supposed to be changed.

findepi commented:
AC & rebased after https://github.com/apache/iceberg/pull/4534 merged.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,77 @@
+    return type;
+  }
+
+  public Set<Integer> columns() {
```
This is the set of fieldIds in the Iceberg table this blob applies to? (Probably should doc that) If these are field ID's is the association between this object and the table made at another level? (I'll probably find out if I keep reading)

findepi replied:
yes and yes

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsCompressionCodec.java:
```diff
@@ -0,0 +1,55 @@
+   * Zstandard single compression frame with content size present
+   */
+  ZSTD("zstd"),
+  /**/;
```
empty JavaDoc

findepi replied:
That's an empty block comment. That's a what we use in Trino to make formatter happy and to allow addition of new enum options without changing lines of existing options (clean git annotate output).

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsCompressionCodec.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  public static StatsCompressionCodec forName(String codecName) {
+    Preconditions.checkNotNull(codecName, "codecName is null");
```
we tend to do these like

https://github.com/apache/iceberg/blob/c75ac359c1de6bf9fd4894b40009c5c42d2fee9d/api/src/main/java/org/apache/iceberg/DistributionMode.java#L51-L54

Where we leverage `valueOf` to throw IllegalArgs for us and check for valid enum names

findepi replied:
https://github.com/apache/iceberg-docs/pull/69 mandates the name to be lowercase.
also, since it's not internal, i wanted to have conversion to and from  file-level name explicit.

rdblue replied:
I agree, we can use the `valueOf` with `toUpperCase`.

findepi replied:
This will be considered a "reference implementation" of the spec. If we're flexible here, we should be flexible in the spec. Otherwise we would add support for situation that's invalid and should never occur, as if mandating it existence.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  static ByteBuffer compressFooterPayload(ByteBuffer payload) {
+    return compress(StatsCompressionCodec.LZ4, payload);
```
Could we make a static final variable for this like FOOTER_COMPRESSION_CODEC, or something like that just to make it clear in the code?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+  private static Compressor getCompressor(StatsCompressionCodec codec) {
+    switch (codec) {
+      case LZ4:
+        // TODO currently not supported
```
Isn't this required for the footers at the moment?

findepi replied:
no, it's optional

rdblue replied:
If this isn't supported, then why list it in the spec? I think having it in the spec implies that it is required.

findepi replied:
Spec says compression is optional.

rdblue replied:
If compression is optional, then why throw `UnsupportedOperationException` when the codec is `NONE`?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+    throw new UnsupportedOperationException("Unsupported codec: " + codec);
+  }
+
+  private static byte[] swap(byte[] bytes) {
```
swap is a bit unclear here since we actually are just reversing the byte order here. Maybe "invert" or "reverse"?

findepi replied:
renaming to `swapBytes`

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+
+  static final int MAGIC_AS_NUMBER_LE = new BigInteger(swap(getMagic())).intValueExact();
+
+  static final int SUPPORTED_FLAGS = 0b1;
```
Should this be 
`SUPPORTED_FLAGS = FLAG_COMPRESSED`
Then when we add more flags it's just
`SUPPORTED_FLAGS = FLAG_COMPRESSED | FLAG_OTHER`

findepi replied:
good idea

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+                decompressed,
+                0,
+                decompressed.length);
+        Preconditions.checkState(decompressedLength == decompressed.length, "Invalid decompressed length");
```
Is this possible? Just wondering if ZstdDecompressor.getDecompressedSize() can return an invalid length. If so should we be throwing a "Corrupted file error or something like that?"

findepi replied:
it cannot return an invalid length

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,177 @@
+      byte[] footer = readInput(fileLength - footerSize, footerSize);
+
+      checkMagic(footer, 0);
+      checkMagic(footer, footerSize - 4);
```
Checking for the magic at the beginning and end of the footer?

RussellSpitzer replied:
Maybe make 4 here a final constant FOOTER_MAGIC_OFFSET

findepi replied:
Obviously there is a bunch of numbers that are hard to follow, and 4 is one of them.
I figured that defining constants for each of them would only blur the code. 
This is a low-level code and i don't think there are any abstractions here.

findepi replied:
added constants for these

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,177 @@
+      checkMagic(footer, 0);
+      checkMagic(footer, footerSize - 4);
+
+      int fileFormatVersion = StatsFormat.readIntegerLittleEndian(footer, footerSize - 8);
```
8 => FOOTER_FILE_FORMAT_OFFSET

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,177 @@
+          fileFormatVersion,
+          StatsFormat.CURRENT_FORMAT_VERSION);
+
+      int flags = StatsFormat.readIntegerLittleEndian(footer, footerSize - 12);
```
FOOTER_FLAG_OFFSET

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,177 @@
+      Preconditions.checkState((flags & ~StatsFormat.SUPPORTED_FLAGS) == 0, "Unsupported flags: %s", flags);
+      boolean compressed = isFlagSet(flags, StatsFormat.FLAG_COMPRESSED);
+
+      int reserved = StatsFormat.readIntegerLittleEndian(footer, footerSize - 16);
```
FOOTER_RESERVED_OFFSET

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,177 @@
+  }
+
+  /**
+   * @throws IOException          when I/O error occurs
```
Should add the full java doc for this as well not just exception stubs, also nit: we don't do vertical alignment usually

findepi replied:
> we don't do vertical alignment usually

That's what the formatter (imported from the definition linked in the project) produced for me, not my own art.

findepi replied:
fixed by not throwing IOException anymore (per suggestion elsewhere)

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,177 @@
+
+  private int getFooterSize() throws IOException {
+    if (!knownFooterSize.isPresent()) {
+      int footerTailSize = Math.toIntExact(Math.min(fileLength, 20));
```
What's the math here? If the file length is less than 20 I thought we can't have a valid footer? (Magic + Reserved + Flag + Format + Magic) * 4 = 20?

findepi replied:
good point

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,177 @@
+      int footerTailSize = Math.toIntExact(Math.min(fileLength, 20));
+      byte[] footerTail = readInput(fileLength - footerTailSize, footerTailSize);
+
+      checkMagic(footerTail, footerTailSize - 4);
```
Like before I think it would be helpful if a lot of these magic numbers had Constants to label them.

findepi replied:
fixed

RussellSpitzer reviewed (commented):
CC: @ggershinsky Could you take a look at this as well I'm not sure how we fit this into our current encryption plans since we are compressing blob blocks within this file seperately we would end up with issues if we encrypted the entire file. Is there an easy way we can compress/encrypt the blobs inside the index files?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,76 @@
+
+  @JsonCreator
+  public BlobMetadata(
+      @JsonProperty("type") String type,
```
What is the benefit of the multiple blobs in a file if we don't have footer information saying why we may or may not need to check a specific blob? I guess currently we can prune them based on "columns" but I feel like maybe we would benefit from other information? Like if a particular file had sketches that applied to different row groups in a file so the correct sketch to use would be based on min and max column values or something ? Just thinking about possible future directions.

findepi replied:
The footer contains a type of every blob. In particular, reading application will read only the blobs of types it can utilize for current operation. Same/similar information will be added to table metadata, so that reading application doesn't need to read stats file footer if no blobs are useful to it.

kbendick commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+        byte[] decompressed =
+            new byte[Math.toIntExact(ZstdDecompressor.getDecompressedSize(inputBytes, inputOffset, inputLength))];
+        int decompressedLength =
+            new ZstdDecompressor().decompress(
```
Is the `ZstdDecompressor` something that we can allocate once and share somehow?

findepi replied:
it's mutable

kbendick commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,141 @@
+        break;
+
+      case ZSTD: {
+        byte[] inputBytes;
+        int inputOffset;
+        int inputLength;
+        if (input.hasArray()) {
+          inputBytes = input.array();
+          inputOffset = input.arrayOffset();
+          inputLength = input.remaining();
+        } else {
+          // TODO implement ZstdDecompressor.getDecompressedSize for ByteBuffer to avoid copying
+          inputBytes = new byte[input.remaining()];
+          input.get(inputBytes);
+          inputOffset = 0;
+          inputLength = inputBytes.length;
+        }
```
You might be interested in `org.apache.iceberg.util.ByteBuffers`, which has a `toByteArray` method that looks very similar to this.

There are other methods in that class you might consider using too, or placing any additional new `ByteBuffer` utility functions.

findepi replied:
ByteBuffers.toByteArray  is awesome, thanks!

ggershinsky commented:
> CC: @ggershinsky Could you take a look at this as well I'm not sure how we fit this into our current encryption plans since we are compressing blob blocks within this file seperately we would end up with issues if we encrypted the entire file. Is there an easy way we can compress/encrypt the blobs inside the index files?

Certainly, will do

findepi commented on build.gradle:
```diff
@@ -217,6 +217,7 @@ project(':iceberg-core') {
       exclude group: 'org.tukaani' // xz compression is not supported
     }
 
+    implementation 'io.airlift:aircompressor'
```
I don't know when a new notice should be added, deferring to committers' judgement

rdblue replied:
This is already included because it is pulled in through ORC.

findepi replied:
Does it mean i should change something here?

rdblue replied:
No, the dependency is already listed in the LICENSE and NOTICE files so there is nothing more to do.

findepi commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,68 @@
+  private final long length;
+  private final String compressionCodec;
+
+  public BlobMetadata(String type, Set<Integer> columns, long offset, long length, @Nullable String compressionCodec) {
```
None is not a compression. The file format spec defines there should be no mention of compression coded, when blob data is not compressed. Internally, in Java code we could still have an enum with additional value, but it would need to require special casing when reading, writing and compressing. Originally i had `Optional` here, but got a comment to use a nullable value instead.

rdblue replied:
I agree with Kyle. I think this would be cleaner if you added `NONE` and always passed the codec as the enum.

findepi replied:
Be your way

findepi commented:
AC

ggershinsky commented:
> CC: @ggershinsky Could you take a look at this as well I'm not sure how we fit this into our current encryption plans since we are compressing blob blocks within this file seperately we would end up with issues if we encrypted the entire file. Is there an easy way we can compress/encrypt the blobs inside the index files?

@RussellSpitzer : had a quick look; I think we should be ok with using the same encrypting stream we plan for other metadata (like manifests). This is not a flat file encryption, won't affect the internal structure, compression etc.

findepi commented:
(just rebased, no other changes)

findepi commented:
Superseded by https://github.com/apache/iceberg/pull/4741

rdblue commented on core/src/main/java/org/apache/iceberg/util/JsonUtil.java:
```diff
@@ -162,7 +162,10 @@ public static Set<Integer> getIntegerSetOrNull(String property, JsonNode node) {
     if (!node.has(property) || node.get(property).isNull()) {
       return null;
     }
+    return getIntegerSet(property, node);
```
Nit: we would normally include an empty newline between the control flow block and the following statement.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,68 @@
+
+  public BlobMetadata(String type, Set<Integer> columns, long offset, long length, @Nullable String compressionCodec) {
+    this.type = Preconditions.checkNotNull(type, "type is null");
+    this.columns = ImmutableSet.copyOf(Preconditions.checkNotNull(columns, "columns is null"));
```
I think columns should probably be a list rather than a set because order matters when computing inputs to blobs. For example, if you have a blob for NDV across two columns you'd need to know how to construct the value tuple that is used as input to the theta sketch.

findepi replied:
Good catch. In fact, blob input should also allow duplicate column inputs too

rdblue commented on core/src/main/java/org/apache/iceberg/stats/FileMetadata.java:
```diff
@@ -0,0 +1,44 @@
+  private final Map<String, String> properties;
+
+  public FileMetadata(List<BlobMetadata> blobs, Map<String, String> properties) {
+    this.blobs = ImmutableList.copyOf(Preconditions.checkNotNull(blobs, "blobs is null"));
```
Style: We generally prefer to keep `Preconditions` first in the method, and not embed them in other statements.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/FileMetadataParser.java:
```diff
@@ -0,0 +1,147 @@
+  private static final String COLUMNS = "columns";
+  private static final String OFFSET = "offset";
+  private static final String LENGTH = "length";
+  private static final String COMPRESSION_CODEC = "compression_codec";
```
As I noted on the spec, I think this should be `compression-codec` to align with the convention used in other JSON parsers.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/FileMetadataParser.java:
```diff
@@ -0,0 +1,147 @@
+    }
+  }
+
+  public static FileMetadata fromJson(JsonNode json) {
```
Is this used elsewhere or can it be package-private?

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsCompressionCodec.java:
```diff
@@ -0,0 +1,55 @@
+    this.codecName = codecName;
+  }
+
+  public String getCodecName() {
```
Style: Iceberg doesn't use `get` in method names. It usually signals that there's either a more specific verb, like `load` or `find` or `fetch`, or is just needless filler. In this case, I'd omit it.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  static int readIntegerLittleEndian(byte[] data, int offset) {
+    return Byte.toUnsignedInt(data[offset]) |
```
Why use `0xFF & value` above, but `Byte.toUnsignedInt` here?

findepi replied:
That's a different conversion, even if syntactically looks similar.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,143 @@
+    return output;
+  }
+
+  private static Compressor getCompressor(StatsCompressionCodec codec) {
```
Style: `get` in a method name.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsCompressionCodec.java:
```diff
@@ -0,0 +1,55 @@
+
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public enum StatsCompressionCodec {
```
Can this be package-private?

findepi replied:
no, that's part of StatsWriter API

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,143 @@
+          inputOffset = 0;
+          inputLength = inputBytes.length;
+        }
+        byte[] decompressed =
```
Style: Iceberg adds empty lines between control flow blocks and the following statements.

rdblue replied:
I guess it doesn't matter since we use the byte array path anyway.

findepi replied:
> Style: Iceberg adds empty lines between control flow blocks and the following statements.

added



> I guess it doesn't matter since we use the byte array path anyway.

i don't understand

rdblue replied:
I think I added the second comment as a follow up on the wrong thread. You can ignore it.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,143 @@
+          inputLength = input.remaining();
+        } else {
+          // TODO implement ZstdDecompressor.getDecompressedSize for ByteBuffer to avoid copying
+          inputBytes = ByteBuffers.toByteArray(input);
```
Is there a maximum size that we could copy out in this case?

findepi replied:
The blob size. We didn't impose any limits on the size of blobs.

rdblue replied:
This is where my other comment was supposed to go. This doesn't matter because the buffer isn't just used to get the size. It is also decompressed so it should be the whole blob.

findepi replied:
This is unclear to me, what should i do here?

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,143 @@
+  private static ByteBuffer compress(StatsCompressionCodec codec, ByteBuffer input) {
+    Compressor compressor = getCompressor(codec);
+    ByteBuffer output = ByteBuffer.allocate(compressor.maxCompressedLength(input.remaining()));
+    compressor.compress(input, output);
```
Does this modify `input`? If it does, then we should pass in `input.duplicate()` instead so that the original buffer is unchanged.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+  private Optional<Integer> knownFooterSize;
+  private Optional<FileMetadata> knownFileMetadata = Optional.empty();
+
+  public StatsReader(InputFile inputFile, Optional<Integer> footerSize) {
```
What is expected for `footerSize`? Should that match `FooterPayloadSize` or be larger by some constant to account for magic and other non-payload bytes?

findepi replied:
size of Footer (in bytes). Footer is defined in the file format spec https://github.com/apache/iceberg-docs/pull/69 
````
### Footer structure

Footer has the following structure

```
Magic FooterPayload FooterPayloadSize Reserved Flags FileFormatVersion Magic
```
````

rdblue replied:
After going through the code, it's clear that this is the footer payload size + a constant. This should be documented in Javadoc.

findepi replied:
yes, this is  footer payload size + a constant, but a user doesn't control "footer payload size" on its own

Footer size can be obtained 

- from manually inspecting the file structure, and `Footer` 's size this is
- from stats writer, as it exposes this information for convenience

I can surely add some javadoc but i don't know what it should contain besides "@param footerSize the size of the file footer defined in the file format specification". Please advise.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+  public StatsReader(InputFile inputFile, Optional<Integer> footerSize) {
+    this.fileLength = inputFile.getLength();
+    this.input = Preconditions.checkNotNull(inputFile, "inputFile is null").newStream();
+    this.knownFooterSize = Preconditions.checkNotNull(footerSize, "footerSize is null");
```
It seems strange to me to check for null with an `Optional`. We typically don't use `Optional` and just pass `empty` as `null`. Since you're checking anyway, that seems like a reasonable update.

rdblue replied:
> to hell with null safety.

What do you mean?

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+  private final long fileLength;
+  private final SeekableInputStream input;
+  private Optional<Integer> knownFooterSize;
+  private Optional<FileMetadata> knownFileMetadata = Optional.empty();
```
In Iceberg, we typically use `null` to signal that a lazy variable hasn't been initialized. That works cleanly with `transient`.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  private static void checkMagic(byte[] data, int offset) {
+    int read = StatsFormat.readIntegerLittleEndian(data, offset);
```
I think converting the magic number to an int makes this code harder to reason about. Magic is not checked in a tight loop, so I don't think that the optimization of comparing the bytes all at once is worth it. I'd prefer to just check each byte in a loop and avoid needing to validate that this is correct:

```java
static final int MAGIC_AS_NUMBER_LE = new BigInteger(swapBytes(getMagic())).intValueExact();
```

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+    return (setFlags & testedFlag) == testedFlag;
+  }
+
+  private int getFooterSize() throws IOException {
```
What do you think about using the same optimization as in Parquet and reading the last 16kb of the file, then reading more if the footer size is larger than that? I think we may as well start with that.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+  public FileMetadata getFileMetadata() throws IOException {
+    if (!knownFileMetadata.isPresent()) {
+      int footerSize = getFooterSize();
+      byte[] footer = readInput(fileLength - footerSize, footerSize);
```
The new `RangeReadable` interface has a `readTail` method that you can use to avoid needing `fileLength`. S3 allows you to read the last N bytes of the file without knowing where to seek to.

findepi replied:
How to i get `RangeReadable` from an `InputFile`?

findepi replied:
I don't do additional HEAD request, because file size and footer size are known. (https://github.com/apache/iceberg-docs/pull/77)
i see that with `RangeReadable.readTail` i could remove the file size and footer size from the table snapshot, which would be a good thing. To do this, i would need to assume some maximum for a "reasonable" footer size and do a `readTail` read with that size (i saw a suggestion to read 16kB, that would work).
It wouldn't change number of requests sent (since currently  file size and footer size are known), but it would slightly reduce amount of data kept in a table.

I assume we don't want to make such an improvement for some file systems while actually degrading performance on other file systems, right? What file systems is RangeReadable available on?

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+      boolean compressed = isFlagSet(flags, StatsFormat.FLAG_COMPRESSED);
+
+      int reserved = StatsFormat.readIntegerLittleEndian(footer, footerSize - 16);
+      Preconditions.checkState(reserved == 0, "Unexpected reserved bytes value: %s", reserved);
```
This breaks forward compatibility if we ever decide to use the reserved bytes, which defeats the purpose. If we have to increment the format version to use them, we should just remove them.

findepi replied:
> This breaks forward compatibility

Yes it does, intentionally. A reader cannot read a file version it doesn't support.
This is same as with unrecognized flags. If there is a flag set a reader doesn't know meaning of, it cannot proceed.

> If we have to increment the format version to use them, we should just remove them.

It exists as reserved so that it's easier to add use of it.
Footer has a fixed-layout portion which is hard to evolve, i.e. hard to maintain support for multiple different layouts in same code base. This is why this field is added from the start.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+            ByteStreams.readFully(input, bytes);
+            ByteBuffer data = ByteBuffer.wrap(bytes);
+            if (blobMetadata.compressionCodec() != null) {
+              StatsCompressionCodec codec = StatsCompressionCodec.forName(blobMetadata.compressionCodec());
```
Why not translate to/from `StatsCompressionCodec` in the parser? That would make more sense to me.

findepi replied:
Forward compatibility for compression codecs. If we add e.g. LZ6, old reader can still make use of the file, it just cannot read portions of the file.

findepi replied:
Yes, I can add `StatsCompressionCodec.UNKOWN`. To me it feels like a special value, very much like `-1`, which I would rather avoid. What would be the benefit from such a change? 

If the user code was interested in compression information, a String is better than an UNKNOWN constant, since String conveys all information, while UNKNOWN already lost something.
If the user code is not dealing with compression information directly, it doesn't matter, it's an impl detail.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+              StatsCompressionCodec codec = StatsCompressionCodec.forName(blobMetadata.compressionCodec());
+              data = StatsFormat.decompressBlob(codec, data);
+            }
+            return new SimpleEntry<>(blobMetadata, data);
```
You can use Iceberg's `Pair` for this, rather than `Map.Entry`.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+   * @throws UncheckedIOException when I/O error occurs
+   */
+  // Note: The method is marked as throwing IOException to allow future implementation evolution.
+  // Currently, the exception isn't being thrown.
```
In Iceberg, we typically omit `IOException` and wrap with `UncheckedIOException` because that has to happen in so many places anyway. I think you can remove the exception.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+   */
+  // Note: The method is marked as throwing IOException to allow future implementation evolution.
+  // Currently, the exception isn't being thrown.
+  public Stream<Map.Entry<BlobMetadata, ByteBuffer>> readAll(List<BlobMetadata> blobs) throws IOException {
```
Why return a `Stream`? Is there a case where you wouldn't read all of the blobs into memory at once?

We also typically use `Iterable` rather than `Stream`.

findepi replied:
> Why return a `Stream`? Is there a case where you wouldn't read all of the blobs into memory at once?

i can imagine such future uses. Like merging stats files.

> We also typically use Iterable rather than Stream.

ok

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,183 @@
+
+  public StatsReader(InputFile inputFile, Optional<Integer> footerSize) {
+    this.fileLength = inputFile.getLength();
+    this.input = Preconditions.checkNotNull(inputFile, "inputFile is null").newStream();
```
You might consider lazily opening the stream.

findepi replied:
I don't see such uses yet. It seems that it's something easy to add later, if needed, and also something that would just clutter the code, if not needed.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,145 @@
+      throw new IllegalStateException(String.format("Property '%s' already set", name));
+    }
+  }
+  public void append(
```
Missing newline between methods.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,145 @@
+  public void append(
+      String type,
+      Set<Integer> columnsCovered,
+      ByteBuffer blobData,
```
`blob`?

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,145 @@
+      String type,
+      Set<Integer> columnsCovered,
+      ByteBuffer blobData,
+      Optional<StatsCompressionCodec> compression) throws IOException {
```
Shouldn't compression be configured for the writer and then applied to every blob uniformly?

findepi replied:
i don't think same compression codec should be used irrespectively of the payload.

i can see a use of default compression codec, but that's something that can be added later easily.

rdblue replied:
Sounds reasonable. I like the idea of a default codec rather than having to pass it in with every blob.

findepi replied:
added

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    outputStream.write(StatsFormat.getMagic());
+    headerWritten = true;
```
When modifying state by setting an instance field, we like to use `this.` so it is obvious that it isn't a local variable.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,145 @@
+
+    Preconditions.checkNotNull(type, "type is null");
+    long fileOffset = outputStream.getPos();
+    ByteBuffer data;
```
It doesn't look like you need a separate declaration anymore.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,145 @@
+    data = compression.map(codec -> StatsFormat.compressBlob(codec, blobData))
+        .orElse(blobData);
+    int length = data.remaining();
+    Channels.newChannel(outputStream).write(data);
```
Looks like channels aren't guaranteed to write all of the bytes, so this should check the return value from `write`.

rdblue replied:
Using a channel is a good idea. You'll just need a `writeFully` method, I think.

findepi replied:
thanks, fixed

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsWriter.java:
```diff
@@ -0,0 +1,145 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class StatsWriter implements Closeable {
```
I think this should use the builder pattern that we use for other file formats. That way all of the metadata and config are set ahead of time and the writer just needs to append each payload. You can see `Parquet` for examples, but it would look like this:

```java
  try (FileAppender<Blob> writer = StatsFiles.write(outputFile)
      .set("key", "value")
      .compressFooter()
      .build()) {
    writer.addAll(blobs);
  }
```

findepi replied:
added

findepi commented:
Reopened per discussion on slack: https://apache-iceberg.slack.com/archives/C02TZ3NMBRC/p1652299433000019

findepi commented:
Thanks @rdblue for a detailed review!

Applied comments except for API change https://github.com/apache/iceberg/pull/4537#discussion_r870916478

alexjo2144 commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,70 @@
+
+public class BlobMetadata {
+  private final String type;
+  private final List<Integer> columns;
```
this is called `columnsCovered` in the Blob class. Maybe unify the field names?

findepi replied:
good catch. in both places it should be "fields"-related. Will go with "inputFields".

alexjo2144 commented on core/src/main/java/org/apache/iceberg/stats/BlobMetadata.java:
```diff
@@ -0,0 +1,70 @@
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+
+public class BlobMetadata {
+  private final String type;
```
Should we have an enum for the Blob types?

findepi replied:
This needs to be forward-compatible and extensible.

alexjo2144 commented on core/src/main/java/org/apache/iceberg/stats/FileMetadata.java:
```diff
@@ -0,0 +1,46 @@
+
+  public FileMetadata(List<BlobMetadata> blobs, Map<String, String> properties) {
+    Preconditions.checkNotNull(blobs, "blobs is null");
+    Preconditions.checkNotNull(properties, "properties is null");
```
The spec file says `properties` is optional, so should it be nullable here?

findepi replied:
`FileMetadataParser` sets empty map when properties are null and doesn't serialize anything if this is empty.

alexjo2144 commented on core/src/main/java/org/apache/iceberg/stats/StatsFiles.java:
```diff
@@ -0,0 +1,68 @@
+    return new WriteBuilder(outputFile);
+  }
+
+  public static class WriteBuilder {
```
Does it buy you anything to have this as a nested class? It's the only thing in the StatsFile class

findepi replied:
Consistency, see eg `org.apache.iceberg.parquet.Parquet.WriteBuilder`

alexjo2144 commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,162 @@
+    private final int bitNumber;
+
+    Flag(int byteNumber, int bitNumber) {
+      this.byteNumber = byteNumber;
```
Is it worth verifying both are positive and bitNumber < 8?

findepi replied:
added

alexjo2144 commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,162 @@
+
+  static void writeIntegerLittleEndian(OutputStream outputStream, int value) throws IOException {
+    outputStream.write(0xFF & value);
+    outputStream.write(0xFF & value >> 8);
```
Maybe worth putting some parenthesis in for clarity

findepi replied:
good idea

alexjo2144 commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,162 @@
+  private static Compressor createCompressor(StatsCompressionCodec codec) {
+    switch (codec) {
+      case NONE:
+        break;
```
Won't think fall through to the `throw UnsupportedOperationException`?

findepi replied:
i don't get it

alexjo2144 replied:
I guess this should never be called with `NONE`, but I'd expect this to return something like a `NoOpCompressor` instead of throwing an UnsupportedException.

findepi replied:
Yes, it's never called with `NONE`, so `NoOpCompressor` would be a dead code.
to make the code clearer, i refactored this a bit

findepi commented:
Thanks @alexjo2144 for your review. AC

rdblue commented on core/src/main/java/org/apache/iceberg/stats/Blob.java:
```diff
@@ -0,0 +1,66 @@
+
+  public Blob(
+      String type, List<Integer> inputFields, ByteBuffer blobData,
+      @Nullable StatsCompressionCodec requestedCompression) {
```
Is this still `@Nullable` now that there is a `NONE` symbol?

findepi replied:
It is. When not set, the writer's configured default compression is applied, per https://github.com/apache/iceberg/pull/4537#discussion_r871523692

rdblue commented on core/src/main/java/org/apache/iceberg/stats/FileMetadataParser.java:
```diff
@@ -0,0 +1,147 @@
+  private static final String PROPERTIES = "properties";
+
+  private static final String TYPE = "type";
+  private static final String COLUMNS = "columns";
```
I think we should rename this in the spec as well to be more clear.

findepi replied:
Fixed. In the spec it's already `fields`

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,165 @@
+    FOOTER_PAYLOAD_COMPRESSED(0, 0),
+    /**/;
+
+    private static final Table<Integer, Integer, Flag> byByteAndBit = Stream.of(values())
```
`static final` variables should be named using ALL_CAPS.

findepi replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,188 @@
+
+  private int footerSize() throws IOException {
+    if (knownFooterSize == null) {
+      if (fileSize < StatsFormat.FOOTER_STRUCT_LENGTH) {
```
Minor: we would normally use `Preconditions.checkState` for this.

findepi replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,188 @@
+
+      int footerPayloadSize = StatsFormat.readIntegerLittleEndian(
+          footerStruct, StatsFormat.FOOTER_STRUCT_PAYLOAD_SIZE_OFFSET);
+      knownFooterSize = 4 /* magic */ + footerPayloadSize + StatsFormat.FOOTER_STRUCT_LENGTH;
```
Rather than "magic" I think it is more clear to say "Footer start magic". I'd also make this a constant that comes from the magic bytes so we don't have to find hard-coded instances.

findepi replied:
done

findepi commented on bundled-guava/src/main/java/org/apache/iceberg/GuavaClasses.java:
```diff
@@ -73,6 +74,7 @@ public class GuavaClasses {
     ImmutableList.class.getName();
     ImmutableMap.class.getName();
     ImmutableSet.class.getName();
+    ImmutableTable.class.getName();
```
Table is a very good API that simplifies use wherever an engineer is tempted to use two-level nested Maps.

Yes, i can go without Table, at the cost of readability. Up to you.

rdblue replied:
I'll take a look at it, this sounds reasonable.

rdblue replied:
I think we can replace this one fairly easily. But it looks like a good idea to add `Resources`.

findepi commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,165 @@
+  }
+
+  static final int FOOTER_MAGIC_OFFSET = 0;
+  static final int FOOTER_STRUCT_LENGTH = 12;
```
fixed-length binary part of Footer.

better name?

findepi commented:
AC

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFiles.java:
```diff
@@ -0,0 +1,68 @@
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public final class StatsFiles {
```
To match the other formats, I think this should just be `Puffin`, as in `Puffin.write(outFile)compressFooter().compressBlobs("zstd").build()`

findepi replied:
good point

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFiles.java:
```diff
@@ -0,0 +1,68 @@
+ * under the License.
+ */
+
+package org.apache.iceberg.stats;
```
What about moving this to `org.apache.iceberg.puffin` or just `org.apache.puffin`?

findepi replied:
Which one you prefer?

findepi replied:
went ahead with `org.apache.iceberg.puffin`.

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsFormat.java:
```diff
@@ -0,0 +1,166 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Table;
+import org.apache.iceberg.util.ByteBuffers;
+
+public final class StatsFormat {
```
`PuffinFormat`? Does this need to be public?

findepi replied:
fixed

rdblue commented on core/src/main/java/org/apache/iceberg/stats/StatsReader.java:
```diff
@@ -0,0 +1,187 @@
+import org.apache.iceberg.stats.StatsFormat.Flag;
+import org.apache.iceberg.util.Pair;
+
+public class StatsReader implements Closeable {
```
`PuffinReader`? and `PuffinWriter`?

findepi replied:
applied!

rdblue commented on core/src/test/java/org/apache/iceberg/stats/StatsFormatTestUtil.java:
```diff
@@ -0,0 +1,44 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.io.ByteStreams;
+
+public final class StatsFormatTestUtil {
```
I think most of what needs to be done is just renaming classes and variables that use "StatsFormat" or "STATS_FORMAT" since it isn't just stats.

findepi replied:
updated

danielcweeks commented on core/src/main/java/org/apache/iceberg/puffin/PuffinFormat.java:
```diff
@@ -0,0 +1,164 @@
+
+  static final PuffinCompressionCodec FOOTER_COMPRESSION_CODEC = PuffinCompressionCodec.LZ4;
+
+  static byte[] getMagic() {
```
Feels like this would be better as a static member than a method since we're creating a new array each time it's called.

findepi replied:
it's shared between classes and array isn't constant, so i didn't want to share array constant.

this is called o(1) times overall though, since each class makes its own trusted copy.

danielcweeks commented on core/src/main/java/org/apache/iceberg/puffin/PuffinReader.java:
```diff
@@ -0,0 +1,187 @@
+
+  public PuffinReader(InputFile inputFile, @Nullable Long fileSize, @Nullable Long footerSize) {
+    Preconditions.checkNotNull(inputFile, "inputFile is null");
+    this.fileSize = fileSize == null ? inputFile.getLength() : fileSize;
```
Seeing as how many of these files are likely going to be small, we might want to consider using `RangeReadable` interface to try to opportunistically read in the whole file with a `readTail` to reduce the number of round-trips to the storage layer.  @findepi this is similar to Trino's `readTail`.

We could just add a default implementation for other FileIO's that mimics the behavior with other calls.

findepi replied:
@danielcweeks thanks for the suggestion. please see my earlier response https://github.com/apache/iceberg/pull/4537#discussion_r872407417 .  does it answer the concerns here?

findepi replied:
added "Statistics are informational. A reader can choose to
ignore statistics information. Statistics support is not required to read the table correctly." in https://github.com/apache/iceberg/pull/4945

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinFormat.java:
```diff
@@ -0,0 +1,164 @@
+    FOOTER_PAYLOAD_COMPRESSED(0, 0),
+    /**/;
+
+    private static final Table<Integer, Integer, Flag> BY_BYTE_AND_BIT = Stream.of(values())
```
Rather than adding a new Guava class, I think you could just use a `Map<Pair<Integer, Integer>, Flag>`. The lookup is hidden behind a static method anyway, so there isn't much of a cost to using `Pair`.

findepi replied:
awesome idea!

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/FileMetadataParser.java:
```diff
@@ -0,0 +1,147 @@
+    try {
+      StringWriter writer = new StringWriter();
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      generator.useDefaultPrettyPrinter();
```
Minor: usually, we add a flag to turn on pretty printing, but leave it off by default.

findepi replied:
thanks!

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinReader.java:
```diff
@@ -0,0 +1,187 @@
+  private Integer knownFooterSize;
+  private FileMetadata knownFileMetadata;
+
+  public PuffinReader(InputFile inputFile, @Nullable Long fileSize, @Nullable Long footerSize) {
```
I don't think this should be public. The reader should be created by helpers in `Puffin`.

findepi replied:
yeah, i agree. i added helper there

rdblue commented on core/src/test/java/org/apache/iceberg/puffin/TestPuffinFormat.java:
```diff
@@ -0,0 +1,63 @@
+import static org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestPuffinFormat {
```
Rather than hard-coding, or at least in addition, I think these tests should validate against the behavior of reading the little-endian value from a ByteBuffer. That will help readers see that it is correct.

findepi replied:
i am not sure i understand, please exemplify

rdblue replied:
I think you should also validate these tests using the standard library. For example:

```java
  byte[] leBytes = bytes(42,0,0,0);
  ByteBuffer.wrap(leBytes).order(ByteOrder.LITTLE_ENDIAN).getInt() == readIntegerLittleEndian(leBytes, 0)
```

rdblue commented on core/src/test/java/org/apache/iceberg/puffin/TestPuffinFormat.java:
```diff
@@ -0,0 +1,63 @@
+  public void testWriteIntegerLittleEndian() throws Exception {
+    testWriteIntegerLittleEndian(0, bytes(0, 0, 0, 0));
+    testWriteIntegerLittleEndian(42, bytes(42, 0, 0, 0));
+    testWriteIntegerLittleEndian(Integer.MAX_VALUE - 5, bytes(0xFa, 0xFF, 0xFF, 0x7F));
```
Nit: lower-case `a` should be upper case.

findepi replied:
right!

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java:
```diff
@@ -0,0 +1,35 @@
+  /**
+   * 8-bytes integer stored little-endian and representing number of distinct values
+   */
+  public static final String NDV_LONG_LITTLE_ENDIAN = "ndv-long-little-endian";
```
This was removed from the spec, so we should also remove it from this commit.

findepi replied:
of course, removed!

findepi commented:
Now that the format spec https://github.com/apache/iceberg/pull/4944 is merged (thank you @rdblue! )
i think this one is ready to merge too as it implements the format.

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,189 @@
+
+  public void finish() throws IOException {
+    checkNotFinished();
+    writeHeaderIfNeeded();
```
Minor: Why write the header? Isn't the file empty if there is no header by this point? We would normally avoid creating the file in that case.

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,189 @@
+  private void writeFooter() throws IOException {
+    FileMetadata fileMetadata = new FileMetadata(writtenBlobsMetadata, properties);
+    ByteBuffer footerJson = ByteBuffer.wrap(FileMetadataParser.toJson(fileMetadata, false)
+        .getBytes(StandardCharsets.UTF_8));
```
Style: it's hard to read whether `.getBytes` is called on the result of `toJson` or `wrap` because of this line wrapping. We generally prefer making it clear by wrapping earlier:

```java
    ByteBuffer footerJson = ByteBuffer.wrap(
        FileMetadataParser.toJson(fileMetadata, false).getBytes(StandardCharsets.UTF_8));
```

findepi replied:
fixed!

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  private void writeFully(ByteBuffer buffer) throws IOException {
+    WritableByteChannel channel = Channels.newChannel(outputStream);
```
I think this needs to be wrapped in try-with-resources because `Channel` is `Closeable`.

findepi replied:
per `Channels.newChannel` docs:

>  Closing the channel will in turn cause the stream to be closed.

this is not what we want here. 

is there a better way to write a `ByteBuffer` to an `OutputStream`?

rdblue replied:
Got it. How about we wrap the stream and keep a reference to the channel instead?

findepi replied:
if it is only about an IDE false positive warning about a Closeable not being closed, i think the right recourse is to add a suppression.

in fact, i would prefer to avoid Channel at all, because Channels interact with interrupts in a manner that's not intuitive to me (eg https://github.com/delta-io/delta/commit/7a3f1e8ec626e80880d524c2b897a969c8b4d63a).

Please let me know which approach would look awesome from Iceberg maintainers' point of view.

rdblue replied:
If you'd like to avoid using Channel, that's fine with me. I just want to make sure we're not leaking resources and this uses a factory method so we aren't guaranteed to have a channel that doesn't require being closed.

Why don't you write a method to implement `writeFully` from a byte buffer in `IOUtil` and we can just call that instead?

findepi replied:
> I just want to make sure we're not leaking resources and this uses a factory method so we aren't guaranteed to have a channel that doesn't require being closed.

right

or, an implementation which is buffering, and we end up not writing (all) the data

> Why don't you write a method to implement writeFully from a byte buffer in IOUtil and we can just call that instead?

will do!

findepi replied:
fixed

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  public long fileSize() {
+    return fileSize.orElseThrow(() -> new IllegalStateException("File not written yet"));
```
`written` -> `closed`?

findepi replied:
no, because it's written (but not closed) in finish

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,189 @@
+  @Override
+  public void add(Blob blob) {
+    Preconditions.checkNotNull(blob, "blob is null");
+    try {
```
`checkNotFinished`?

findepi replied:
added, thanks

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/Puffin.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    public PuffinWriter build() {
+      return new PuffinWriter(outputFile, properties, compressFooter, defaultBlobCompression);
```
Could you set `created-by` here?

findepi replied:
added a dedicated setter to set it in properties

findepi replied:
How would you envision that on the Builder's API level?

rdblue commented on core/src/test/java/org/apache/iceberg/puffin/TestPuffinReader.java:
```diff
@@ -0,0 +1,112 @@
+    Puffin.ReadBuilder readBuilder = Puffin.read(inputFile)
+        .withFileSize(inputFile.getLength());
+    if (footerSize != null) {
+      readBuilder = readBuilder.withFooterSize(footerSize);
```
Is it worth testing the cases where the footer size is wrong?

findepi replied:
good idea, added

rdblue commented on core/src/test/java/org/apache/iceberg/puffin/TestPuffinWriter.java:
```diff
@@ -0,0 +1,124 @@
+    assertThatThrownBy(writer::footerSize)
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessage("Footer not written yet");
+    assertThatThrownBy(writer::finish)
```
Why allow an explicit call to `finish`?

findepi replied:
so that a user can read file size and footer size from the writer object.
otherwise it wouldn't be possible to use writer with try-with-resources

rdblue reviewed (approved):
This looks good to me. The only blocker is closing the byte channel.

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/Puffin.java:
```diff
@@ -0,0 +1,97 @@
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public final class Puffin {
```
Adding Javadoc to this class and the builders would be really helpful. Especially since footer compression is currently broken!

findepi commented:
Applied comments. @rdblue can you please give it another look?

findepi commented:
Rebased after https://github.com/apache/iceberg/pull/5019 merged.

nastra commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  private void checkNotFinished() {
+    if (finished) {
```
nit: Preconditions.checkState(..) maybe?

nastra commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,190 @@
+  public void finish() throws IOException {
+    checkNotFinished();
+    writeHeaderIfNeeded();
+    if (footerSize.isPresent()) {
```
nit: Preconditions.checkState(..) maybe?

findepi replied:
thanks for the suggestion!

nastra commented on core/src/test/java/org/apache/iceberg/puffin/TestFileMetadataParser.java:
```diff
@@ -0,0 +1,120 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestFileMetadataParser {
```
maybe add a few tests with invalid/missing json fields?

findepi replied:
good idea.

this let me realize there is a problem with trailing contentn

i.e. this should fail:
```java
FileMetadataParser.fromJson("{\"blobs\": []} {}")
```

but it does not.

I think i am following the established JSON parsing patterns, so it probably is a pre-existing problem

findepi replied:
> I think i am following the established JSON parsing patterns, so it probably is a pre-existing problem

https://github.com/apache/iceberg/pull/5048

nastra commented on core/src/test/java/org/apache/iceberg/puffin/TestPuffinFormat.java:
```diff
@@ -0,0 +1,86 @@
+    buffer.flip();
+    byte[] written = new byte[4];
+    buffer.get(written);
+    Preconditions.checkState(Arrays.equals(written, expected), "Invalid expected value");
```
should that be just a `assertThat(written).isEqualTo(expected)`? Because this will actually show expected/actual values if the check fails, whereas the precondition doesn't show that

findepi replied:
Throwing an AssertionError here could be interpreted as a bug in the implementation under test.
i used assertions for the code under test.
i use checkState to verify the test is configured properly. 

> Because this will actually show expected/actual values if the check fails, whereas the precondition doesn't show that

agreed, there is a cost to it

nastra commented on core/src/test/java/org/apache/iceberg/puffin/TestPuffinFormat.java:
```diff
@@ -0,0 +1,86 @@
+
+  private void testReadIntegerLittleEndian(byte[] input, int offset, int expected) {
+    // Sanity check: validate the expectation
+    Preconditions.checkArgument(
```
same here, should this be just using `assertThat(...).isEqualTo(..)`?

nastra commented on core/src/main/java/org/apache/iceberg/puffin/PuffinReader.java:
```diff
@@ -0,0 +1,187 @@
+    this.input = inputFile.newStream();
+    if (footerSize != null) {
+      Preconditions.checkArgument(0 < footerSize && footerSize <= this.fileSize - MAGIC.length,
+          "Invalid footer size: %s", footerSize);
```
should this maybe mention what the valid size for the footer is?

findepi replied:
well, at this point we don't know what the valid footer size is.
we only know the provided one is _obviously_ not valid

nastra commented on core/src/test/java/org/apache/iceberg/puffin/TestPuffinWriter.java:
```diff
@@ -0,0 +1,124 @@
+        .hasMessage("Unsupported codec: LZ4");
+    assertThatThrownBy(writer::close)
+        .isInstanceOf(UnsupportedOperationException.class)
+        .hasMessage("Unsupported codec: LZ4");
```
I'm not sure this is intuitive for users to understand why this exception is happening (given that currently LZ4 is the default when using `compressFooter()` and it's not implemented). Should the default compressor maybe be configured to ZSTD?

nastra reviewed (commented):
overall LGTM, just a few suggestions here and there

findepi commented on core/src/main/java/org/apache/iceberg/puffin/BlobMetadata.java:
```diff
@@ -0,0 +1,81 @@
+    return type;
+  }
+
+  public List<Integer> inputFields() {
```
Iceberg field IDs are integers, so the implementation is chosen to be limited to integers.

We can maybe change `fields            | list of JSON long` in the puffin spec to be list of integers.

rdblue replied:
Yes, we should update the spec to match the type used by the table spec for field IDs.

findepi replied:
https://github.com/apache/iceberg/pull/5087

findepi commented on core/src/main/java/org/apache/iceberg/puffin/PuffinFormat.java:
```diff
@@ -0,0 +1,166 @@
+    Flag(int byteNumber, int bitNumber) {
+      Preconditions.checkArgument(
+          0 <= byteNumber && byteNumber < PuffinFormat.FOOTER_STRUCT_FLAGS_LENGTH,
+          "Invalid byteNumber");
```
This is enum initialization code. It can fail only if an engineer changes the enum definition itself and then they will be able to quickly realize what the problem is.

findepi commented on core/src/main/java/org/apache/iceberg/puffin/PuffinFormat.java:
```diff
@@ -0,0 +1,166 @@
+      case ZSTD:
+        return compress(new ZstdCompressor(), input);
+    }
+    throw new UnsupportedOperationException("Unsupported codec: " + codec);
```
i think both are fine

findepi commented on core/src/main/java/org/apache/iceberg/puffin/PuffinFormat.java:
```diff
@@ -0,0 +1,166 @@
+        break;
+
+      case ZSTD: {
+        byte[] inputBytes;
```
good idea, moving

findepi commented:
Thanks @rdblue and @nastra for your reviews!

Updated the code accordingly. Please take another look.

rdblue commented on core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java:
```diff
@@ -0,0 +1,181 @@
+    ByteBuffer footerPayload = PuffinFormat.compress(footerCompression, footerJson);
+    outputStream.write(MAGIC);
+    int footerPayloadLength = footerPayload.remaining();
+    writeFully(footerPayload);
```
Minor: why not call `IOUtil.writeFully(outputStream, footerPayload)` directly rather than having a private `writeFully` method that just adds the output stream?

findepi replied:
https://github.com/apache/iceberg/pull/5088

rdblue commented on core/src/main/java/org/apache/iceberg/io/IOUtil.java:
```diff
@@ -46,6 +50,21 @@ public static void readFully(InputStream stream, byte[] bytes, int offset, int l
+    if (!buffer.hasRemaining()) {
+      return;
+    }
+    byte[] chunk = new byte[WRITE_CHUNK_SIZE];
```
Rather than allocating every time this is called, can you create a `ThreadLocal` to share this buffer? Alternatively, you could pass the temporary buffer in.

findepi replied:
> Alternatively, you could pass the temporary buffer in.

This poses sizing challenge. I.e. the caller needs to provide a reasonably sized buffer.

> Rather than allocating every time this is called, can you create a ThreadLocal to share this buffer?

Sure, this is feasible. Do you happen to know what would be the expected reuse ratio for such a buffer?

Alternatively we can have a static buffer pool that lends buffers to the current thread.
(assuming we have a problem that we want to fix here)

rdblue commented:
Thanks, @findepi! This looks good to me. We can still follow up, but I think the majority of the changes are ready so I've merged it to avoid keeping a big PR outstanding.
